contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+interval. The shorter the measurement interval, the better the quality of
+recommendations for spiky, latency sensitive workloads. Minimum reasonable
+resolution is 1/min, recommended is 1/sec.
+
+For memory the objective is to **keep the probability of the container usage
+exceeding the request in a specific time window below a certain threshold**
+(e.g. below 1% in 24h). The window must be long (≥ 24h) to ensure that evictions
+caused by OOM do not visibly affect (a) availability of serving applications
+(b) progress of batch computations (a more advanced model could allow user to
+specify SLO to control this).
+
+#### Handling OOMs ####
+When a container is evicted due to exceeding available memory, its actual memory
+requirements are not known (the amount consumed obviously gives the lower
+bound). This is modelled by translating OOM events to artificial memory usage
+samples by applying a "safety margin" multiplier to the last observed usage.
+
+### History Storage ###
+VPA defines data access API for providers of historical events and resource
+utilization. Initially we will use InfluxDB as the reference implementation of
+this API. Users will be able to plug their own implementations.
+
+History Storage is populated with real time updates of resources utilization and
+events from Heapster, similarly to the Recommender. The storage keeps at least
+8 days of data. This data is only used to initialize the Recommender on startup.
+
+### Open questions ###
+1. How to resolve conflicts if multiple VPA objects match a Pod.
+
+2. What happens to VPA API objects, if for any reason admin decides to disable
+   VPA in the cluster.
+
+-----------------------
+Alternatives considered
+-----------------------
+
+### Pods point at VPA ###
+*REJECTED BECAUSE IT REQUIRES MODIFYING THE POD SPEC*
+
+#### proposal: ####
+Instead of VPA using label selectors, Pod Spec is extended with an optional
+field `verticalPodAutoscalerPolicy`,
+a [reference](https://kubernetes.io/docs/api-reference/v1/definitions/#_v1_localobjectreference)
+to the VPA config.
+
+#### pros: ####
+* Consistency is enforced at the API level:
+    * At most one VPA can point to a given Pod.
+    * It is always clear at admission stage whether the Pod should use
+      VPA or not. No race conditions.
+* It is cheap to find the VPA for a given Pod.
+
+#### cons: ####
+* Requires changing the core part of the API (Pod Spec).
+
+### VPA points at Deployment ###
+
+#### proposal: ####
+VPA has a reference to Deployment object. Doesn’t use label selector to match
+Pods.
+
+#### pros: ####
+* More consistent with HPA.
+
+#### cons: ####
+* Extending VPA support from Deployment to other abstractions that  manage Pods
+  requires additional work. VPA must be aware of all such abstractions.
+* It is not possible to do a rolling update of the VPA config.
+  For example setting `max_memory` in the VPA config will apply to the whole
```

Comment:
The VPA could probably take as input on its spec a label with which to subdivide pods under its selector.  For deployment, that would be the generation/hash label.  For stateful sets, that could be region.